atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/jmx/CassandraJmxCompactionManager.java

```diff
@@ -37,111 +39,65 @@
  */
 public class CassandraJmxCompactionManager {
     private static final Logger log = LoggerFactory.getLogger(CassandraJmxCompactionManager.class);
-    private final Set<CassandraJmxCompactionClient> clients;
+    private final ImmutableSet<CassandraJmxCompactionClient> clients;
     private final ExecutorService exec;
 
-    private CassandraJmxCompactionManager(Set<CassandraJmxCompactionClient> clients, ExecutorService exec) {
-        this.clients = clients;
-        this.exec = exec;
+    private CassandraJmxCompactionManager(ImmutableSet<CassandraJmxCompactionClient> clients, ExecutorService exec) {
+        this.clients = Preconditions.checkNotNull(clients);
+        this.exec = Preconditions.checkNotNull(exec);
     }
 
-    public static CassandraJmxCompactionManager newInstance(Set<CassandraJmxCompactionClient> clients, ExecutorService exec) {
-        Preconditions.checkNotNull(clients);
-        Preconditions.checkNotNull(exec);
+    public static CassandraJmxCompactionManager create(ImmutableSet<CassandraJmxCompactionClient> clients, ExecutorService exec) {
         return new CassandraJmxCompactionManager(clients, exec);
     }
 
-    public Set<CassandraJmxCompactionClient> getCompactionClients() {
-        return Collections.unmodifiableSet(clients);
+    public void performTombstoneCompaction(long timeoutInSeconds,
+                                           String keyspace,
+                                           String tableName) throws InterruptedException, TimeoutException {
+        Stopwatch stopWatch = Stopwatch.createStarted();
+        removeHintedHandoff();
+        long remainingTimeoutSeconds = timeoutInSeconds - stopWatch.elapsed(TimeUnit.SECONDS);
+
+        // ALL hinted handoffs need to be deleted before moving to tombstone compaction task
+        deleteTombStone(keyspace, tableName, remainingTimeoutSeconds);
     }
 
-    /**
-     *
-     * @param timeoutInSeconds - timeout for compaction.
-     * @param keyspace - keyspace for the tables to be compacted
-     * @param tableName - tables to be compacted
-     * @throws TimeoutException - TimeoutException is thrown when compaction cannot finish in a given time period.
-     * @throws InterruptedException
-     */
-    public void performTombstoneCompaction(final long timeoutInSeconds,
-                                     final String keyspace,
-                                     final String tableName) throws TimeoutException, InterruptedException {
-        if(clients.isEmpty()) {
-            log.error("No compaction client can be found.");
+    private void removeHintedHandoff() throws InterruptedException {
+        CassandraJmxCompactionClient client = Iterables.get(clients, 0);
+        try {
+            client.truncateAllHints();
```

Comment:
I read the source code in Cassandra that backs this feature; "truncateAllHints" doesn't actually mean truncate all hints, it means truncate all hints just for that node (it just issues a CF.truncate on that node's local system keyspace's hints CF).

You will still need to iterate through multiple nodes to tell them to all do this, similar to how the logic here used to be.